state.go

```diff
@@ -397,6 +397,16 @@ func NewUpdateErrorState(err menderError, update UpdateResponse) State {
 func (ue *UpdateErrorState) Handle(c Controller) (State, bool) {
 	// TODO error handling
 	c.ReportUpdateStatus(ue.update, statusFailure)
+
+	// TODO fetch logs from logger
+	c.UploadLog(ue.update, []LogEntry{
+		LogEntry{
+			Timestamp: time.Now().Format(time.RFC3339),
+			Level:     "error",
+			Message:   "update failed",
+		},
+	})
```

Comment:
You're right, this might deserve to be broken down even further:

* Fetch failure -> UpdateErrorState1
* Install failure -> UpdateErrorState2
* Boot update fails -> rollback & reboot -> UpdateErrorState3
* Boot update succeeds, but reporting update fails -> rollback & reboot -> UpdateErrorState4

The numbered `UpdateErrorStates` are arbitrary, just for discussion. The two first ones are as you say, harmless, nothing has really happened to the device yet. However, IMHO I think the states 2 and higher should be treated the same, since they are unlikely to be temporary, and should be reported in a mandatory fashion. The first `UpdateErrorState1` could however go back to `UpdateCheck`, since being unable to fetch is probably due to network outage. But I guess this one is debatable.

In any case, for such a "softer" type of error we could introduce a new state `UpdateSoftErrorState` or something similar, which unlike `UpdateErrorState`, can go straight back to `UpdateCheck`, even if the report fails.

And yes, we probably need to store some additional state for this.